portico: Add integrated feature comparison table.

karlstolley opened the pull request:
This PR prepares an integrated feature-comparison table for Cloud and Self-Hosted features, which can also be compared side-by-side.

Rather than maintaining separate tables, this unifies all feature information into a single table for easier maintenance and feature parity.

The `IntersectionObserver` Web API is used to better display individual feature-comparison header rows when in the "stuck" position, without causing any meaningful change to how the feature-comparison rows appear within the body of the table.

[Prototype](https://terpimost.github.io/zulip-comparison/) by @terpimost 

**Screenshots**

_Firefox screen captures do not respect the blur on the Zulip menu, but it is there--despite these screenshots._

| Cloud | Self-hosted | All |
| --- | --- | --- |
| ![cloud-plans](https://github.com/zulip/zulip/assets/170719/ea0199c6-9347-4f67-9c67-acb59930e609) | ![self-hosted-plans](https://github.com/zulip/zulip/assets/170719/6839f729-64dc-4f04-8b80-dccf5167c1c9) | ![all-plans](https://github.com/zulip/zulip/assets/170719/8b8c8ca3-20e8-4e76-a3cc-1e5367e26d4f) |

| All, medium | All, small | All, x-small |
| --- | --- | --- |
| ![all-plans-full](https://github.com/zulip/zulip/assets/170719/5215eb19-2e6c-485b-ade9-ba5d4d0ca20d) | ![all-plans-medium](https://github.com/zulip/zulip/assets/170719/4debaac5-133b-4708-adbb-880643ec9dff) | ![all-plans small](https://github.com/zulip/zulip/assets/170719/12bf8cbf-c9ac-4604-bf03-577b8337823c) |

**All Plans scrolling behavior**

![all-plans-sticky-scrolling](https://github.com/zulip/zulip/assets/170719/7c3caea4-eb0a-410c-ada9-41e5983f3bca)




<details>
<summary>Draft-stage notes</summary>

* The integrated table can be inspected on the Cloud tab on http://localhost:9991/plans
* Missing/null values are currently represented by red "N/A" (the red is temporary, to make them easier to spot).

![Screen Shot 2024-01-31 at 19 02 52](https://github.com/zulip/zulip/assets/170719/07e28b83-0d13-4902-bf38-9620ecae0d45)

A few content discrepancies need resolving (as of 2024-01-31):

- [ ] Sorting may need to be rethought based on feature availability (e.g., in the Security section)
- [ ] “Configurable call provider” has BigBlueButton on Self-hosted, but not Cloud
- [ ] “TLS encryption with Let’s Encrypt” on Self-hosted, but just “TLS encryption” on Cloud
- [ ] “GDPR compliant deployment” on Self-hosted, “GDPR compliant” on Cloud
- [ ] “Import data from another Zulip server” only listed on Cloud, not Self-hosted

</details>

alya commented:
> A few content discrepancies need resolving (as of 2024-01-31):

By the way, this list was super helpful!

alya commented:
Testing this PR:
- The Cloud plans comparison heading is missing when you navigate there from the self-hosted tab.

<img width="691" alt="Screenshot 2024-02-05 at 4 08 18 PM" src="https://github.com/zulip/zulip/assets/2090066/ba507a3e-e7d5-42ff-b9e1-7373c5fa1509">

- Let's drop the "same 100% open-source software" line from the Cloud tab (matching what we currently have); I think it's more confusing than helpful there.

alya commented:
The prototype doesn't have a hover effect on the selected tab, and I think that looks better.

alya commented:
I also noticed that the "And [much more](http://localhost:9991/plans/#self-hosted-plan-comparison)!" link from the other pages (/for/business, /self-hosting) doesn't work to go to the feature table anymore; it just takes me to the top of /plans.

karlstolley commented:
Okay @alya -- I've moved this out of draft status, as it's now feature-complete. I've also addressed your feedback above:

* Headings display as expected when using the buttons
* Hover effects are corrected on the buttons
* The Cloud table has neither the open-source software line nor the legend
* The "And much more!" links from other pages should now work

alya commented:
Looks great to me!

alya commented:
While we're here, let's make this "Limited" yellow. I think that's more correct in terms of what the colors mean, and it would be consistent with the yellow "Limited" we have elswewhere.

<img width="833" alt="Screenshot 2024-02-07 at 11 29 29 PM" src="https://github.com/zulip/zulip/assets/2090066/26f88d1d-f160-4617-8453-668fe1d660dd">

alya commented:
Looking over the data, the only tweak I noticed was:

- Azure Active Directory should be a checkmark on Standard. [see [here](https://chat.zulip.org/#narrow/stream/107-kandra/topic/cloud.20plans.3A.20Azure.20AD/near/1730414)]

karlstolley commented:
Great! Pushed both of those fixes:

Yellow on Limited (tooltip, not pictured, takes the yellow, too):
![Screenshot 2024-02-08 at 9 22 14 AM](https://github.com/zulip/zulip/assets/170719/c7e782ef-8211-43d4-93ee-ec9991c6e841)

Green checkmark on Cloud standard for Azure SSO:
![Screenshot 2024-02-08 at 9 22 46 AM](https://github.com/zulip/zulip/assets/170719/82c7aad2-3aa2-4bf1-a00c-02c06504b623)

timabbott commented on templates/corporate/comparison_table_integrated.html:
```diff
@@ -488,7 +491,7 @@ <h2>Self-hosted plans comparison</h2>
             <tr>
                 <td class="comparison-table-feature">
-                    <a href="/help/bots-and-integrations">
+                    <a href="/help/bots-overview">
```
This hunk probably belongs in the previous commit.

karlstolley replied:
Oops, yeah. I resolved that change from the old files in the wrong spot. I'll fix that.

timabbott commented:
Looks really nice! I managed to get into this weird state on /plans where I had some higher-up headers stuck together (scrolling up and down after getting into this state, they stay glued with the overlap).

![image](https://github.com/zulip/zulip/assets/2746074/c370d912-9834-423f-95df-ecb72601edd2)

I don't have a reproducer but I did manage to make it happen within 30s 3 times in a row by just scrolling up and down with my mouse's scroll wheel in various patterns (fast seemed to not do it; it seemed more going upwards past "INTEGRATIONS" was usually the point where it happened, but I'm not certain).

karlstolley commented:
Got it. Do you happen to know roughly how large your viewport was? Given that there's blue space to the left, I'm guessing this was pretty large.

There are some brittle adjustments I had to make in the CSS to avoid those kinds of states, so having that information will probably lead to a fix for this. But I'll investigate.

timabbott commented:
The window was full screen on a 1920x1080 monitor.

karlstolley commented:
Thanks, @timabbott. I was able to get into this state a few times by quickly whipping the scrollbar up and down. What seems to happen is that sudden movements can sometimes cause elements in the "stuck" position, which ought to take a `stuck` class from `IntersectionObserver`, to be missing that class--causing them to appear higher in the header row than they should (so it's not the element lower in the row that's the problem, in other words).

I have lowered the `threshold` value in the Intersection Observer ops from `1.0` (100% of the element in the target intersection area) to `0.8` (80% in the target area). I think the reduced requirement for intersecting pixels give `IntersectionObserver` a bit more wiggle room to work with for more liberally assigning the `stuck` class, but it would be worth your trying to reproduce with this new setting. (Note that `IntersectionObserver` fires asynchronously, so while that makes it very performant, it's also not out of the realm of possibility that sudden scrolling movements could produce this effect.)

I've also pushed the history here so that the `/help/bots-overview` link is fixed in the commit previous to the one noted above.

timabbott commented:
I can't reproduce that issue anymore :). I think this seems like a good checkpoint to merge. I did observe that it's possible to get a state where at certain scroll positions it looks similar, but it wouldn't stick. I wonder if changing the `z-index` to have the stuck one on top of the other would provide the more expected visual appearance:

![image](https://github.com/zulip/zulip/assets/2746074/04ba5d13-27c9-4cab-b16b-093d53712387)

But given it's a rare scroll position, easily fixed by scrolling, I don't think it worth spending more than a few minutes of attention on.

Merged, thanks for all the work on this @karlstolley!

karlstolley commented:
Oh boy. I think that was premature. I didn't update the conversation here, but that fix I was hopeful about didn't seem to work. I'll put up a patch here in a bit.
